README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# aMSE

<!-- badges: start -->
<!-- badges: end -->

This packages up a new Abalone Management Strategy Evaluation framework.
It has a novel structure that is based around the spatial ideas of
populations within spatial management units (SAUs), within a zone:

- zone - highest geographical level. Is simply the totality of the
  spatial management units.

- SAU - spatial assessment units. In Tasmania these would currently be
  the classical statistical blocks.

- population - literally a population. These are the active components
  within the simulation framework. The dynamics of the simulation are
  based around the populations, although, with positive larval dispersal
  (the default) there is some dependency of neighbouring populations on
  each other.

This conceptual structure has implications for how the simulations are
conditioned during the management strategy testing. While it is still
the case that the general properties of each SAU are used to set the
scene for each SAU’s component populations, the advent of sufficient GPS
logger data now allows a more formal definition of the productivity of
each population. Currently, the modelling assumes that each population
is linearly arranged around the coastline, and this will allow, where
available, specific catch/yield data to be used to define the bounds of
each population within an SAU.

An important change from previous designs is that larval dispersal is
implemented explicitly rather than implicitly being held constant. This
alters the dynamics so that analytical equilibrium methods no longer
work and we need to resort to iterative approaches to equilibrium if the
larval dispersal rate is \> 0.0.

Of course, adding such a component increases the number of options that
may need to be explored, but currently we envisage including a very low,
a middle range, and a relatively high level of larval dispersal, to
examine its implications.

# LATEST UPDATES

- 2023-07-10 aMSE 0.1.6 Modified TasHS and aMSE so that instead of using
  hsargs$startCE and hsargs$endCE to define a vector of reference years
  when using the constantrefhcr in Tasmania, one should now use a
  hsargs\$refperiodCE which should be a vector of years describing the
  constant reference period to be used (assuming one is using the
  constantrefhcr). Within TasHS there is also a checkhsargs() function
  which provides a brief check of the hsargs being used (strictly that
  is still under development).

- 2023-06-11 aMSE 0.1.5 Modified the condition plot and fixed up a few
  places where numbers were hard wired into the software. Introduced a
  new option to save output following projections (just in case of
  experiments with subsequent analyses crash - I’m looking at you SA).
  Now one can set interimout =c(outdir,postfixdir) and that should save
  a tempout into your outdir directory. See the help for do_MSE.

- 2023-06-04 aMSE 0.1.4 Small modifications to generalize across
  jurisdictions and allow for the consthcr constant catch harvest
  strategy; also removed irrelavent messages when plotting the phase
  plot

- 2023-06-03 aMSE 0.1.3 Modified to allow the mcdahcr and constrefhcr
  HCR to run successfully with the conditioned SA setup.

- 2023-04-17 aMSE 0.1.2 Amended save_hsargs so it can handle data.frame
  within hsargs.

- 2023-04-12 aMSE 0.1.1 Started generalizing the outputs from aMSE by
  beginning to separate functions that relate to specific jurisdiction
  outputs into a separate R package (in Tasmania it is currently called
  aMSEExtra).

- 2023-04-04 aMSE 0.1.0 Added curryear to hcrfun so all HS functions
  should now include curryear inthier arguments. Within doprojecitons()
  it is set to year inside the loop. This was needed to allow the SA HS
  to use its fleet dynamics model and is not needed in Tasmania, but is
  required in the argument list.

- 2023-03-31 aMSE 0.0.25 Modified calcpopC througout to allow for SA’s
  harvest strategy and fleet dynamics.

- 2022-12-14 aMSE 0.0.24 Lots of minor edits and modifications aimed at
  improving the comparison of scenarios.

- 2022-11-11 aMSE 0.0.23 Added a new tab = recruits, containing plots of
  the stock recruitment relationships at an sau and at a population
  within sau level. Added a table to the Tables tab.

- 2022-11-10 aMSE 0.0.22 now have the capacity to introduce productivity
  events in specific projection years. This can now be done at sau
  specific levels. One defines the survivorship of recruits in each
  selected year and also the survivorship of all post-settlement
  animals. No changes are needed to one’s control file if this option is
  not being used.

- 2022-11-02 aMSE 0.0.21 added changes to how do_MSE responds to hsargs.

- 2022-10-31 aMSE 0.0.20 added do_comp_outputs, scenebyvar, scenebyzone,
  projectiononly, and sauquantbyscene to assist with comparing
  scenarios.

## Installation

Once the MSE framework is made public you can install the development
version from [GitHub](https://github.com/haddonm/aMSE) with:

``` r
if (!require(devtools)){install.packages("devtools")} 

devtools::install_github("https://github.com/haddonm/aMSE",build_vignettes = TRUE)
```

Alternatively, while the development version remains private, you can
generate a branch that you can work on by cloning the repository, which,
again, can be done very simply within RStudio. Open the New Project
option in the project dialog at the top right of the RStudio screen and
selection Version Control, then use ‘<https://github.com/haddonm/aMSE>’
in the top box, identify where you want the new directory put, and press
return. ALternatively, you could download the zip file from inside the
‘code’ button and establish an R project from that.

It would be a good idea to read Hadley Wickham’s draft chapter on Git
and GitHub at <https://r-pkgs.org/index.html>.

## Example

This is an example which illustrates the generation of an initial
equilibrium, which then goes on to apply an early version of the
Tasmania MCDA using only 50 replicates (that bit takes about 15 seconds
on my computer, hence 2.5 minutes for 1000). It uses built in data-sets
but usually you would read in a control file, which would contain the
name of the biological datafile describing each population.

``` r
# a constant TAC example
starttime <- (Sys.time())
options("show.signif.stars"=FALSE,
        "stringsAsFactors"=FALSE,
        "max.print"=50000,
        "width"=240)
# declare libraries ------------------------------------------------------------
suppressPackageStartupMessages({
  library(aMSE)
  library(TasHS)
  library(hutils)
  library(hplot)
  library(makehtml)
  library(knitr)
})
dbdir <- getDBdir()
# Obviously you should modify the rundir and datadir to suit your own setup
prefixdir <- paste0(dbdir,"A_CodeUse/aMSEUse/scenarios/")
postfixdir <- "M15h75"
rundir <- paste0(prefixdir,postfixdir)
confirmdir(rundir)
controlfile <- "controlM15h75.csv"
# equilibrium zone -------------------------------------------------------------
# You now need to ensure that there is, at least, a control.csv, and a 
# constantsdata.csv file in the data directory plus some other data .csv files
# depending on how conditioned you want the model to be. Templates for the
# correct format can be produced using ctrlfiletemplate(), datafiletemplate().
# 
# Of course, usually one would use data files, control.csv and a saudata.csv, which
# is listed as the datafile within the control.csv. These must be stored in 
# rundir. There are example files in the DropBox folder:
# .\..\Dropbox\National abalone MSE\aMSE_files\scenarios\examplerun. Read the 
# 'Using_aMSE' section in the Documentation .docx file in the documentation 
# directry  Then, assuming you have the very latest version of aMSE = 500, 
# the code in the documentation file (or in the 'run_aMSE.R' file within the
# examplerun subdirectory should work. 

# HarvestStrategy.R should also be in rundir
source(paste0(rundir,"/TasHS1_Tas.R"))
# run the scenario --------------------- Obviously unhash this to make it work
#
# out <- do_MSE(rundir,controlfile,datadir,hsargs=hsargs,hcrfun=mcdahcr,
#               sampleCE=tasCPUE,sampleFIS=tasFIS,sampleNaS=tasNaS,
#               getdata=tasdata,calcpopC=calcexpectpopC,makeouthcr=makeouthcr,
#               varyrs=7,startyr=48,verbose=TRUE,
#               ndiagprojs=4,savesauout=TRUE)
# 
# makeoutput(out,rundir,datadir,postfixdir,controlfile,openfile=TRUE,verbose=FALSE)
```

After running the whole, even if you do not generate the results
webpage, you could also try:

``` r
#str(out$zoneDP,max.level=1)
#str1(out$hcrout)
```

To see the structure of the dynamic object generated by the projections.

See the vignette Running_aMSE.Rmd for a more detailed example.

See the New.Rd for recent developments


NEWS.md
## aMSE Recent Activity.

See the ReadMe for the latest few events

aMSE v0.0.19 (Release date: 2022-10-26)
==============

Changes:

* added more functions to facilitate comparing scenarios.


aMSE v0.0.18 (Release date: 2022-10-18)
==============

Changes:

* In the projections replicates across years has now been changed to years across replicates to allow for inclusion of FIS performance measures. Also, further comparison functions included.

* 2022-10-26 aMSE 0.0.19 
* 2022-10-18 aMSE 0.0.18 In the projections replicates across years has now been changed to years across replicates to allow for inclusion of FIS performance measures. Also, further comparison functions included.

* 2022-09-26 aMSE 0.0.17 Added finalcondyeardepletion to aid in comparisons.

* 2022-08-24 aMSE 0.0.16 Started to add depensation and other refinements.

* 2022-08-19 aMSE 0.0.15 Added do_comparison as a wrapper function for comparing scenarios.

* 2022-08-05 aMSE 0.0.14 Added more functions to assist with comparing scenarios. 

* 2022-08-05 aMSE 0.0.12 Added more functions to assist with comparing scenarios. Still sorting out references to other packages.

* 2022-08-01 aMSE 0.0.11 Added a bunch of functions to assist with comparing scenarios. No EWN.

* 2022-07-10 aMSE 0.0.10 Switched dependency from hutils to codeutils. re-added zone1 for S21 to the data sets. No EWN.

* 2022-07-10 aMSE 0.0.9 Added data sets to allow for examples and documentation to illustrate the conditioning and running of the MSE. Added adjustavrec as a function, nade a wrapper rnormz and rlnormz to enable zero variation while still using up a random number so the sequence was maintained. Ready for production work. No EWN.  

* 2022-06-29 aMSE 0.0.8 Corrected a large mistake where the estimated SAU harvest rates were being biased high because the SAU's end of year exploitable biomass was being used instead of the mid-year exploitable biomass. 

* 2022-06-22 aMSE 0.0.7 Now have rewritecontrolfile and rewritesaudatafile and the option to include a table of parameters from sizemod into the inputs of aMSE. Have begun development of functions to aid comparing the output from different scenarios, including tables and plots. No EWN.

* 2022-06-08 aMSE 0.0.6 Have revised the ctrltemplate and datatemplate generators so that they reflect a lambda = 0.75 and correctly use the hyperstability. No EWN

* 2022-06-06 aMSE 0.0.5. Have reviewed the dynamics in detail so that the outputs from the aMSE for each SAU more closely match the outputs from sizemod. The original disjunction was partly due to a faulty implementation of the inclusion of the option of hyper-stability, but that has now been amended and the operatnig model is now behaving as it should. Breaking the dynamics within each SAU into multiple sub-populations with variation also affects the equivalence between aMSE and sizemod butmostly influences the productivity rather than the recruitment dynamics and the trajectories mainly follow those from sizemod but with the cpue and biomasses being projected incorrectly. It appears (still working on a why) that by changing the productivity and sub-dividing that productivity among sub-populations, the overall recruitment levels required to mimic the dynamics in all cases appears to be less in aMSE than in sizemod. Otheriwse nothing now needs to change. One uses parameters obtained from sizemod, especially the recruitment deviates, and then apply the functions that search for more optimal values for AvRec, the unfished recruitment levles in each SAU. The fits obtained should be close to those found in sizemod. In 0.0.5 there remain issues relating to the internal data sets, which include one called 'zone', which is not designed to work with the option of hyper-stability (lambda != 1.0). Otherwise, No EWN.

* 2022-05-31 aMSE 0.0.4 Added functions that assist in conditioning the operating model by optimizing the fit to CPUE and size-composition data after opimizing the AvRec values for each SAU. No EWN.

* 2021-11-29 aMSE 0.0.1 Removed datadir. Started implementing hyper-stability while keeping the interface the same, so that the results from sizemod can be used.

* 2021-11-23 aMSE 0.0.0.1 Added functions to simplify the optimization of AvRec relative to the CPUE data after initial parameter estimation for each SAU using the sizemod package. Also added a comparison of size-composition data to the do_condition function. Next step will be to include the optimization of the recdevs using both CPUE and size-comp data. 

* 2021-11-21 aMSE 0.0.0.020 So as to use any observed size-composition data in the conditioning, a number of functions have been added to convert the predicted numbers-at-size in the catch by population into NAS x SAU. These can now be plotted against the observed NAS for each SAU and this can be either to the console or to a new tab in the webpage called 'predictedcatchN'. This work derived from the progress made with the sizemod package used to fit size-based assessment models to individual SAU. Because the MSE sub-divides each SAU into a number of populations each with somewhat different MaxDL, L95, etc, even after fitting the models, when using that data it is then necessary to refit the AvRec values, which, because of the sub-division into populations each having different biology can lead to the MSE AvRec being either smaller or larger than the estimated R0 for a SAU. Similar arguments apply to teh recruitment residuals but generally the pattern provides a reasonable fit.

* 2021-10-17 aMSE 0.0.0.010 As cleanslate, within setuphtml() has been deprecated in the package makehtml, I have removed reference to it throughout aMSE. Now, to clean a directory of aMSE files use cleanrundir.

* 2021-09-22 aMSE 0.0.0.025 Modifications to do_condition and other functions to facilitate the ease of changing values within an array of control files and saudata files.


R/domse.R
# Tas context -----------------
# postfixdir <- "BCconstref"
# rundir <- rundir
# controlfile=controlfile
# hsargs=hsargs
# hcrfun=constantrefhcr  #mcdahcr  # consthcr   #constantrefhcr
# sampleCE=tasCPUE
# sampleFIS=tasFIS
# sampleNaS=tasNaS
# getdata=tasdata
# calcpopC=calcexpectpopC
# varyrs=7
# startyr=38
# verbose=TRUE
# ndiagprojs=4
# makehcrout=makeoutconst   #makeouthcr
# HSPMs=getcpueHS
# fleetdyn=NULL
# cutcatchN=56
# matureL = c(70,200)
# wtatL = c(80,200)
# mincount=100
# includeNAS=FALSE
# depensate=0
# pmwtSwitch = 0
# stablewts = c(0.4, 0.5, 0.1)
# hcrname="constantrefhcr"   #"consthcr"
# hcrscoreoutputs = extractandplotscores
# kobeRP = c(0.4,0.2,0.15)
# interimout=""

# SA context -------------
# rundir=rundir
# controlfile=controlfile
# hsargs=hsargs
# hcrfun=SA_HS
# sampleCE=saCPUE
# sampleFIS=saFIS
# sampleNaS=saNaS
# getdata=sadata
# calcpopC=sacalcexpectpopC
# makeouthcr=samakeouthcr
# makehcrout=makeouthcr
# hcrscoreoutputs = saextractandplotscores
# HSPMs=sagetHSscores
# fleetdyn = safleetdyn
# varyrs=7
# startyr=38
# verbose=TRUE
# ndiagprojs=3
# cutcatchN=56
# matureL=c(40,170)
# wtatL=c(50,200)
# mincount=120
# includeNAS = TRUE
# depensate=0
# kobeRP=c(0.4,0.2,0.15)
# hcrfun=SA_HS
# sampleCE=saCPUE
# sampleFIS=saFIS
# sampleNaS=saNaS
# getdata=sadata
#


#' @title do_MSE an encapsulating function to hold the details of a single run
#'
#' @description do_MSE is a form of meta-function that contains the code and
#'     function calls that constitute a single scenario run for aMSE. This is
#'     the reason the argument list is as long as it is. The code has been
#'     encapsulated like this to simplify development, maintenance, and
#'     distribution. Already with the number of variant conditioning examples
#'     produced as tests of the code-base, errors have arisen through a failure
#'     to propagate all changes to the code now in this function to all the
#'     separate MSE run files used to manage each scenario. A potential flaw
#'     lies with the need to apply the jurisdictionHS functions to any
#'     historical fishery data used to condition the operating model, ready for
#'     the first year of the projections. Currently, the code used is unique to
#'     the Tasmanian case and this obviously still requires generalization.
#'     This may require a new function to be produced in each jurisdictionHS.R
#'     file, but that is still under consideration. Part of the development of
#'     this meta-function will be to further articulate the generation of
#'     results, and the inclusion of error traps on the argument list entries.
#'     To aid in the conditioning the option of not calculating the productivity
#'     has been added. By default it will occur and when projecting it will also
#'     always occur. HSstats is a list that is always saved and contains
#'     performance statistics for the HS used (currently contains sum5 and sum10
#'     the cumulative catches to 5 and 10 years into the projections).
#'
#' @param rundir the full path to the directory in which all files relating to a
#'     particular run are to be held.
#' @param controlfile the filename of the control file present in rundir
#'     containing information regarding the particular run.
#' @param hsargs the constants used to define the workings of the hcr
#' @param hcrfun the name of the harvest control rule that is used to
#'     calculate the aspirational catches by SAU or TAC by zone for the
#'     following year
#' @param sampleCE a function from jurisdictionHS.R that generates the CPUE
#'     statistics from the projected data
#' @param sampleFIS a function from jurisdictionHS.R that generates the FIS
#'     statistics
#' @param sampleNaS a function from jurisdictionHS.R that generates the
#'     Numbers-at-size samples
#' @param getdata a function that gathers all the data required by the hcrfun
#'     and combines it into an hcrdata object ready for the hcrfun
#' @param calcpopC a function that takes the output from hcrfun (either the
#'     aspirational catch x SAU or TAC x zone) and generates the actual catch
#'     per population in each SAU expected in the current year.
#' @param makeouthcr Another JurisdictionHS.R function that generates (or not)
#'     an object that is continually updated by the hcrfun. To avoid very
#'     inefficient code this object (at least for Tasmania) is cycled through
#'     the iterations.
#' @param hcrscoreoutputs A function defined in an extra aMSE package or
#'     source file, that defines how the hcr scores should be dealt with. In
#'     Tasmania they recreated from the projections, plotted and their values
#'     snet out into the scores object.
#' @param HSPMs Another function defined in the HS package that reconstructs
#'     the cpue HSPM scores, and other measures ready for plotting
#' @param fleetdyn a function that calculates the distribution of catch across
#'     the sau and populations. Currently not needed by Tas but needed by SA
#' @param interimout should results be saved after projections have finished?
#'     the default="", which means nothing is saved if no risk of crashing in
#'     subsequent analyses. But if trying new stuff which may waste lots of time
#'     if the projections need to be repeated, the set interimout = c(outdir,
#'     postfixdir), which should be defined in the run_aMSE you are using,
#'     and that will save the main objects needed for later analysis in the
#'     directory used to save final results. The ordering of
#' @param varyrs how many years at the end of the conditioning on the fishery,
#'     data into zoneDD, to which to add recruitment variation, default = 7,
#'     which suits the Tasmanian west coast. Used in prepareprojection
#' @param startyr the index for the first year of the conditioning dynamics to
#'     include in the plotted cpue trajectories, to give startyr:indexoflastyear.
#'     used in sauplots.
#' @param verbose Should progress comments be printed to console, default=FALSE
#' @param ndiagprojs the number of replicate trajectories to plot in the
#'     diagnostics tab to illustrate ndiagprojs trajectories to ensure that
#'     such projections appear realistic; default=3
#' @param cutcatchN to reduce the size of the final array of numbers-at-size
#'     in the catch one can remove all the empty cells below a given size
#'     class. In the default there are 105 2mm size classes and setting
#'     cutcatchN = 56 removes size classes 1:55 2 - 110 mm; only from catchN
#' @param matureL is a vector of 2, default = c(70,200), that define the x-axis
#'     bounds on the biology maturity-at-Length plots
#' @param wtatL is a vector of 2, default = c(80,200), that define the x-axis
#'     bounds on the biology weight-at-Length plots
#' @param mincount given size-composition data what minimum sample size will
#'     be deemed acceptable for inclusion in the plots and conditioning.
#'     default=100.
#' @param includeNAS should the NAS projections be included in the output. If
#'     they are then the final output size is greatly increased. default=FALSE
#' @param depensate should depensation occur. If > 0, eg depensate=0.2, then
#'     depensation will occur when depletion is < whatever value one puts in for
#'     depensate. Implemented inside the function 'oneyearrec'
#' @param kobeRP reference points for the eHS pseudo kobe plot. A vector of the
#'     targdepl, default=0.4, limdepl, default=0.2, and the limH, default=0.15
#'
#' @seealso{
#'  \link{makeequilzone}, \link{dohistoricC}, \link{prepareprojection},
#'  \link{doprojections}, \link{getprojyrC}, \link{oneyearrec}
#' }
#'
#' @return a large list containing tottime, projtime, starttime, glb, ctrl,
#'     zoneDD, zoneDP, projC, condC, sauout, and outzone
#' @export
#'
#' @examples
#' print("wait on suitable data sets in data")
do_MSE <- function(rundir,controlfile,hsargs,hcrfun,sampleCE,sampleFIS,
                   sampleNaS,getdata,calcpopC,makeouthcr,hcrscoreoutputs,
                   HSPMs,fleetdyn=NULL,interimout="",
                   varyrs=7,startyr=42,verbose=FALSE,ndiagprojs=3,
                   cutcatchN=56,matureL=c(70,200),wtatL=c(80,200),mincount=100,
                   includeNAS=FALSE,depensate=0,kobeRP=c(0.4,0.2,0.15)) {
  # generate equilibrium zone -----------------------------------------------
  starttime <- (Sys.time())
  zone <- makeequilzone(rundir,controlfile,verbose=verbose)
  equiltime <- (Sys.time()); if (verbose) print(equiltime - starttime)
  # declare main objects ----------------------------------------------------
  glb <- zone$glb
  ctrl <- zone$ctrl
  zoneC <- zone$zoneC
  zoneD <- zone$zoneD
  zone1 <- zone$zone1
  saudat <- zone$saudat
  constants <- zone$constants
  production <- zone$product
  projC <- zone$zone1$projC
  condC <- zone$zone1$condC
  # save some equil results -------------------------------------------------
  biology_plots(rundir, glb, zoneC, matL=matureL,Lwt=wtatL)
  sauprod <- plotproductivity(rundir,production,glb)
  numbersatsize(rundir, glb, zoneD)
  #Condition on Fishery -----------------------------------------------------
  if (any(condC$initdepl < 1)) {
    initdepl <- condC$initdepl
    if (verbose) cat("Conducting initial depletions  ",initdepl,"\n")
    zoneD <- depleteSAU(zoneC,zoneD,glb,initdepl=initdepl,production)
  }
  if (verbose) cat("Conditioning on the Fishery data  \n")
  zoneDD <- dohistoricC(zoneD,zoneC,glob=glb,condC,calcpopC=calcpopC,
                        fleetdyn=fleetdyn,hsargs=hsargs,sigR=1e-08,sigB=1e-08)
  hyrs <- glb$hyrs
  propD <- as.data.frame(t(getzoneprops(zoneC,zoneDD,glb,year=hyrs)))
  columns <- c("B0","MSY","MSYDepl","bLML","propprot","SpBDepl","catch",
               "harvestR")
  plotpopprops(x=propD,rundir=rundir,glb=glb,varnames=columns,startyr=hyrs,
               bins=21,console=FALSE)
  propD[,"SAU"] <- c(glb$sauname[glb$sauindex],NA)
  addtable(propD,"propertyDD.csv",rundir,category="zoneDD",caption=
             "Properties of zoneD after conditioning on historical catches.")
  addtable(round(t(zoneDD$harvestR[(hyrs-9):hyrs,]),4),"final_harvestR.csv",
           rundir,category="zoneDD",
           caption="Last ten years of population vs harvest rate.")
  addtable(round(saudat,5),"saudat.csv",rundir,category="zoneDD",
           caption="SAU constant definitions")
  popdefs <- as.data.frame(t(getlistvar(zoneC,"popdef")))
  popdefs[,c(1:6,8:18)] <- round(popdefs[,c(1:6,8:18)],3) #so SAU can be txt
  popdefs[,"SAU"] <- glb$sauname[glb$sauindex]  #so SAU can be txt
  addtable(popdefs,"popdefs.csv",rundir,category="zoneDD",
           caption="Population vs Operating model parameter definitions")
  condout <- plotconditioning(zoneDD,glb,zoneC,condC$histCE,
                              histCatch=condC$histCatch,rundir,
                              condC$recdevs,console=FALSE)
  saurecdevs(condC$recdevs,glb,rundir,filen="saurecdevs.png")
  # plot predicted size-comp of catch vs observed size-comps
  catchN <- zoneDD$catchN
  sauCt <- popNAStosau(catchN,glb)
  compdat <- condC$compdat$lfs
  if (!is.null(compdat)) {
    for (plotsau in 1:glb$nSAU) {
      lfs <- preparesizecomp(compdat[,,plotsau],mincount=mincount)
      yrsize <- as.numeric(colnames(lfs))
      histyr <- condC$histyr
      pickyr <- match(yrsize,histyr[,"year"])
      LML <- histyr[pickyr,"histLML"]
      plotsizecomp(rundir=rundir,incomp=lfs,SAU=glb$saunames[plotsau],lml=LML,
                   catchN=sauCt[,,plotsau],start=NA,proportion=TRUE,
                   console=FALSE)
    }
  }
  # do projections ------------------------------------------------------------
  if (verbose) cat("Preparing for the projections \n")
  outpp <- prepareprojection(projC=projC,condC=condC,zoneC=zoneC,glb=glb,
                             calcpopC=calcpopC,zoneDD=zoneDD,
                             ctrl=ctrl,varyrs=varyrs,lastsigR = ctrl$withsigR)
  zoneDP <- outpp$zoneDP
  projC <- outpp$projC
  zoneCP <- outpp$zoneCP
 # Rprof()  #  ctrl$reps=25
  if (verbose) cat("Doing the projections \n")
  outproj <- doprojections(ctrl=ctrl,zoneDP=zoneDP,zoneCP=zoneCP,glb=glb,
                           hcrfun=hcrfun,hsargs=hsargs,sampleCE=sampleCE,
                           sampleFIS=sampleFIS,sampleNaS=sampleNaS,
                           getdata=getdata,calcpopC=calcpopC,
                           makehcrout=makeouthcr,fleetdyn=fleetdyn,verbose=TRUE)

#  Rprof(NULL)
  if (verbose) cat("Now generating final plots and tables \n")
  zoneDP=outproj$zoneDP
  hcrout <- outproj$hcrout; #str(hcrout)
  NAS <- list(Nt=zoneDP$Nt,catchN=zoneDP$catchN)
  zoneDP <- zoneDP[-c(17,16,15)]  # This removes the Nt etc from zoneDP
  if (length(interimout) == 2) {
    outfile <- filenametopath(interimout[1],paste0("temp",interimout[2],".Rdata"))
    if (includeNAS) {
      postprojout <- list(zoneCP=zoneCP,hcrout=hcrout,zoneDP=zoneDP,glb=glb,
                         ctrl=ctrl,condC=condC,constants=constants,NAS=NAS)
    } else {
      postprojout <- list(zoneCP=zoneCP,hcrout=hcrout,zoneDP=zoneDP,glb=glb,
                         ctrl=ctrl,condC=condC,constants=constants)
    }
    save(postprojout,file=outfile)
  }
 # histCE <- condC$histCE
  B0 <- getvar(zoneC,"B0")
  ExB0 <- getvar(zoneC,"ExB0")
  sauout <- sauplots(zoneDP,NAS,glb,rundir,B0,ExB0,
                     startyr=startyr,addCI=TRUE,histCE=condC$histCE)
  diagnosticsproj(sauout,glb,rundir,nrep=ndiagprojs)
  outzone <- poptozone(zoneDP,NAS,glb,
                       B0=sum(getvar(zoneC,"B0")),
                       ExB0=sum(getvar(zoneC,"ExB0")))
  zonesummary <- plotZone(outzone,rundir,glb,startyr=startyr,
                          CIprobs=c(0.05,0.5,0.95),addfile=TRUE)
  fishery_plots(rundir=rundir,glb=glb,select=zoneCP[[1]]$Select,
                histyr=condC$histyr,projLML=projC$projLML)
  historicalplots(rundir=rundir,condC=condC,glb=glb)
  NAS$catchN <- NAS$catchN[(cutcatchN:glb$Nclass),,,]
  projtime <- Sys.time()
  tottime <- round((projtime - starttime),3)
  # calculate HS performance statistics
  sum5 <- getprojyrC(catsau=zoneDP$catsau,glb=glb,period=5)
  sum10 <- getprojyrC(catsau=zoneDP$catsau,glb=glb)
  HSstats <- list(sum10=sum10,sum5=sum5)
  save(HSstats,file=paste0(rundir,"/HSstats.RData"))
  save(glb,file=paste0(rundir,"/glb.RData"))
  save_hsargs(rundir,hsargs)   # prints hsargs to HSPerfs tab
  if (verbose) cat("HSstats.RData, glb.RData, and hsargs.txt saved to rundir \n")
  plothsstats(rundir,HSstats,glb,average=FALSE)
  plothsstats(rundir,HSstats,glb,average=TRUE)
  addtable(hcrout$refpts,"hcrout_refpts.csv",rundir,category="HSperf",
           caption="HCR reference points")
  scores <- hcrscoreoutputs(rundir=rundir,HSPMs=HSPMs,hcrout=hcrout,
                            cpue=sauout$cpue,catches=sauout$catch,
                            glb=glb,yearCE=condC$yearCE,hsargs=hsargs)
  # generate sau phase plots
  nSAU <- glb$nSAU
  kobedata <- vector(mode="list",length=nSAU)
  names(kobedata) <- glb$saunames
  for (plotsau in 1:glb$nSAU) {
    kobedata[[plotsau]] <- HSphaseplot(dyn=sauout,glb=glb,sau=plotsau,
                                       rundir=rundir,startyr=condC$yearCE[1],
                                       console=FALSE,targdepl=kobeRP[1],
                                       limdepl=kobeRP[2],limH=kobeRP[3])
  }

  if (!includeNAS) NAS=NULL
  out <- list(tottime=tottime,projtime=projtime,starttime=starttime,glb=glb,
              ctrl=ctrl,zoneCP=zoneCP,zoneD=zoneD,zoneDD=zoneDD,zoneDP=zoneDP,
              NAS=NAS,projC=projC,condC=condC,sauout=sauout,outzone=outzone,
              hcrout=hcrout,production=production,condout=condout,
              HSstats=HSstats,saudat=saudat,constants=constants,hsargs=hsargs,
              sauprod=sauprod,scores=scores,zonesummary=zonesummary,
              kobedata=kobedata)
  return(out)
} # end of do_MSE








